lib/SPIRV/SPIRVToOCL12.cpp

```diff
@@ -114,6 +138,91 @@ void SPIRVToOCL12::visitCallSPIRVControlBarrier(CallInst *CI) {
       &Attrs);
 }
 
+Instruction *SPIRVToOCL12::visitCallSPIRVAtomicIncDec(CallInst *CI, Op OC) {
+  AttributeList Attrs = CI->getCalledFunction()->getAttributes();
+  return mutateCallInstOCL(
+      M, CI,
+      [=](CallInst *, std::vector<Value *> &Args) {
+        Args.resize(1);
+        return OCLSPIRVBuiltinMap::rmap(OC);
+      },
+      &Attrs);
+}
+
+CallInst *SPIRVToOCL12::mutateCommonAtomicArguments(CallInst *CI, Op OC) {
+  assert(CI->getCalledFunction() && "Unexpected indirect call");
+  AttributeList Attrs = CI->getCalledFunction()->getAttributes();
+
+  return mutateCallInstOCL(
+      M, CI,
+      [=](CallInst *, std::vector<Value *> &Args) {
+        auto Ptr = findFirstPtr(Args);
+        auto NumOrder = getSPIRVAtomicBuiltinNumMemoryOrderArgs(OC);
+        auto ArgsToRemove = NumOrder + 1; // OpenCL1.2 builtins does not use
+                                          // scope and memory order arguments
+        auto StartIdx = Ptr + 1;
+        auto StopIdx = StartIdx + ArgsToRemove;
+        Args.erase(Args.begin() + StartIdx, Args.begin() + StopIdx);
+        return CI->getName();
+      },
+      &Attrs);
+}
+
+Instruction *SPIRVToOCL12::visitCallSPIRVAtomicUMinUMax(CallInst *CI, Op OC) {
+  AttributeList Attrs = CI->getCalledFunction()->getAttributes();
+  return mutateCallInstOCL(
+      M, CI,
+      [=](CallInst *, std::vector<Value *> &Args) {
+        return OCL12SPIRVBuiltinMap::rmap(OC == OpAtomicUMin ? OpAtomicSMin
+                                                             : OpAtomicSMax);
+      },
+      &Attrs);
+}
+
+Instruction *SPIRVToOCL12::visitCallSPIRVAtomicCmpExchg(CallInst *CI, Op OC) {
+  AttributeList Attrs = CI->getCalledFunction()->getAttributes();
+  return mutateCallInstOCL(
+      M, CI,
+      [=](CallInst *, std::vector<Value *> &Args) {
+        // SPIRV OpAtomicCompareExchange and OpAtomicCompareExchangeWeak
+        // has Value and Comparator in different order than ocl functions
+        // both of them are translated into atomic_cmpxchg
+        std::swap(Args[1], Args[2]);
+        return OCL12SPIRVBuiltinMap::rmap(OpAtomicCompareExchange);
+      },
+      &Attrs);
+}
+
+Instruction *SPIRVToOCL12::visitCallSPIRVAtomicBuiltin(CallInst *CI, Op OC) {
+  CallInst *CIG = mutateCommonAtomicArguments(CI, OC);
+
+  Instruction *NewCI = nullptr;
+  switch (OC) {
```

Comment:
For OpAtomicLoad, OpAtomicStore, OpAtomicFlagTestAndSet, OpAtomicFlagClear see my https://github.com/KhronosGroup/SPIRV-LLVM-Translator/issues/266#issuecomment-514263083 and https://github.com/KhronosGroup/SPIRV-LLVM-Translator/issues/266#issuecomment-514579228

Reply:
Support for those four instructions added. Thank you for your comments in #266. I've modified just implementation for OpAtomicFlagTestAndSet, as we don't need to issue icmp here. The only thing we need to do is truncate the i32 result from atomic_xchg to i1, as atomic_xchg returns the old value.

Let's assume old value is 1, then 1 is casted to boolaen true value which means the same as the result of OpAtomicFlagTestAndSet (The instruction’s result is true if the flag was in the set state).
Analogical for the false value.
It's safer than issuing icmp.